File: CS/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraGrid.Views.Grid;

namespace DisabledCells {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e) {
            FillDataSource();
        }

        private void FillDataSource() {
            dataTable1.Rows.Add("Product 1002", "US", true);
            dataTable1.Rows.Add("Product 1001", "Germany", false);
            dataTable1.Rows.Add("Product 1003", "Canada", true);
        }


        private bool IsShipToUSCanada(GridView view, int row) {
            try {
                string val = Convert.ToString(view.GetRowCellValue(row, "ShipCountry"));
                return (val == "US" || val == "Canada");
            }
            catch {
                return false;
            }
        }

        private void gridView1_ShowingEditor(object sender, CancelEventArgs e) {
            if(gridView1.FocusedColumn.FieldName == "IsFreeShipping"
                && IsShipToUSCanada(gridView1, gridView1.FocusedRowHandle))
                e.Cancel = true;
        }

        private void gridView1_RowCellStyle(object sender, DevExpress.XtraGrid.Views.Grid.RowCellStyleEventArgs e) {
            if(e.Column.FieldName == "IsFreeShipping"
                && IsShipToUSCanada(gridView1, e.RowHandle)) {
                e.Appearance.BackColor = Color.LightGray;
            }
        }
    }
}

File: VB/Form1.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo

Namespace DisabledCells

    Public Partial Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
            AddHandler gridView1.CustomDrawCell, AddressOf GridView1_CustomDrawCell
        End Sub

        Private Sub GridView1_CustomDrawCell(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.RowCellCustomDrawEventArgs)
            If Equals(e.Column.FieldName, "gridColumn1") AndAlso IsShipToUSCanada(gridView1, e.RowHandle) Then
                Dim viewInfo As CheckEditViewInfo = TryCast(CType(e.Cell, GridCellInfo).ViewInfo, CheckEditViewInfo)
                viewInfo.CheckInfo.State = DevExpress.Utils.Drawing.ObjectState.Disabled
            End If
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            FillDataSource()
        End Sub

        Private Sub FillDataSource()
            dataTable1.Rows.Add("Product 1002", "US", True)
            dataTable1.Rows.Add("Product 1001", "Germany", False)
            dataTable1.Rows.Add("Product 1003", "Canada", True)
        End Sub

        Private Function IsShipToUSCanada(ByVal view As GridView, ByVal row As Integer) As Boolean
            Try
                Dim val As String = Convert.ToString(view.GetRowCellValue(row, "ShipCountry"))
                Return Equals(val, "US") OrElse Equals(val, "Canada")
            Catch
                Return False
            End Try
        End Function

        Private Sub gridView1_ShowingEditor(ByVal sender As Object, ByVal e As CancelEventArgs)
            If Equals(gridView1.FocusedColumn.FieldName, "IsFreeShipping") AndAlso IsShipToUSCanada(gridView1, gridView1.FocusedRowHandle) Then e.Cancel = True
        End Sub

        Private Sub gridView1_RowCellStyle(ByVal sender As Object, ByVal e As RowCellStyleEventArgs)
            If Equals(e.Column.FieldName, "IsFreeShipping") AndAlso IsShipToUSCanada(gridView1, e.RowHandle) Then
                e.Appearance.BackColor = Color.LightGray
            End If
        End Sub
    End Class
End Namespace
